src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq_vmr.c

```diff
@@ -1259,29 +1260,29 @@ static int xgq_firewall_op(struct platform_device *pdev, enum xgq_cmd_log_page_t
 	ret = (cmd->xgq_cmd_rcode == -ETIME || cmd->xgq_cmd_rcode == -EINVAL) ?
 		0 : cmd->xgq_cmd_rcode;
 
-	if (ret) {
-		struct xgq_cmd_cq_log_page_payload *log = NULL;
-		u32 log_size = 0;
-
-		log = (struct xgq_cmd_cq_log_page_payload *)&cmd->xgq_cmd_cq_payload;
-		log_size = log->count;
-
-		if (log_size > len) {
-			XGQ_ERR(xgq, "return log size %d is greater than request %d",
-				log->count, len);
-			log_size = len;
-		} else if (log_size  == 0) {
-			XGQ_ERR(xgq, "no error message");
-		} else {
-			char *log_msg = vmalloc(log_size);
-			if (log_msg == NULL) {
-				XGQ_ERR(xgq, "vmalloc failed, no msg");
-				goto done;
-			}
-			memcpy_from_device(xgq, address, log_msg, log_size);
-			XGQ_ERR(xgq, "%s", log_msg);
-			vfree(log_msg);
+	/*
+	 * No matter ret is 0 or non-zero, the device might return
+	 * error messages to print into the dmesg.
+	 */
+	log = (struct xgq_cmd_cq_log_page_payload *)&cmd->xgq_cmd_cq_payload;
+	log_size = log->count;
+
+	if (log_size > len) {
+		XGQ_WARN(xgq, "return log size %d is greater than request %d",
+			log->count, len);
+		/* reset to valid shared memory size */
+		log_size = len;
+	}
+
+	if (log_size != 0) {
+		char *log_msg = vmalloc(log_size);
```

Comment:
do you want to do vmalloc(log_size + 1) and log_msg[log_size] = '\0' before mem copy to make sure the string is always 0-terminated?